source/extensions/filters/http/oauth/filter.cc

```diff
@@ -0,0 +1,475 @@
+    auto scheme = Http::Headers::get().SchemeValues.Https;
+
+    const auto* scheme_header = headers.Scheme();
+    if ((scheme_header != nullptr &&
+         scheme_header->value().getStringView() == Http::Headers::get().SchemeValues.Http)) {
+      scheme = Http::Headers::get().SchemeValues.Http;
+    }
+
+    const std::string base_path = absl::StrCat(scheme, "://", host_);
+    const std::string callback_path = base_path + config_->callbackPath();
+    const std::string state_path = absl::StrCat(base_path, headers.Path()->value().getStringView());
+
+    const std::string escaped_redirect_uri =
+        absl::StrReplaceAll(callback_path, escapedReplacements());
+    const std::string escaped_state = absl::StrReplaceAll(state_path, escapedReplacements());
+
+    const std::string new_url =
+        fmt::format(authClusterUriParameters(), config_->oauthServerHostname(), config_->clientId(),
+                    escaped_redirect_uri, escaped_state);
+    response_headers->setReferenceKey(Http::Headers::get().Location, new_url);
+    decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  // At this point, we *are* on /_oauth. We believe this request comes from the authorization
+  // server and we expect the query strings to contain the information required to get the access
+  // token
+  Http::Utility::QueryParams query_parameters = Http::Utility::parseQueryString(path_str);
+  if (query_parameters.find(queryParamsError()) != query_parameters.end()) {
+    sendUnauthorizedResponse();
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  // if the data we need is not present on the URL, stop execution
+  if (query_parameters.find(queryParamsCode()) == query_parameters.end() ||
+      query_parameters.find(queryParamsState()) == query_parameters.end()) {
+    sendUnauthorizedResponse();
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  auth_code_ = query_parameters.at(queryParamsCode());
+  state_ = absl::StrReplaceAll(query_parameters.at(queryParamsState()), unescapedReplacements());
+
+  Http::Utility::Url state_url;
+  if (!state_url.initialize(state_, false)) {
+    sendUnauthorizedResponse();
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+  const auto scheme = state_url.scheme();
+
+  const std::string cb_url = absl::StrCat(scheme, "://", host_, config_->callbackPath());
+  oauth_client_->asyncGetAccessToken(auth_code_, config_->clientId(), config_->clientSecret(),
+                                     cb_url);
+
+  // pause while we await the next step from the OAuth server
+  return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+}
+
+// Always sanitize these legacy Oauth headers that may be maliciously injected to mess with
+// Jenkins's built-in authentication. These will be readded later in the filter after
```

Comment:
id leave out references to specific implementations, we can keep this vague

Reply:
this entire method should go